### PR TITLE
Connect CA Roots

### PR DESCRIPTION
**Pull request by IvanKolchanov**
# Changes made
1. Added GET /v1/agent/connect/ca/roots endpoint
2. Included new endpoint into ICatalogEndpoint.cs
3. Added possible test to AgentTest.cs
4. Edited test_config.json for the test

# Issue ticket number and link
*  Refers to issue https://github.com/G-Research/consuldotnet/issues/331

### CODE REVIEW

**Comment by IvanKolchanov** [comment]
The only issue I have is with another test case that the Consul in Golang has: https://github.com/hashicorp/consul/blob/v1.9.17/api/connect_ca_test.go#L12 .
They use the testutil to create another client for testing and do not bootstrap connect in that test case, which is, however, not really possible with the current testing setup that consuldotnet has. Because that would require to create another test_config.json and run both the new and the old one when testing. Please correct me if I am wrong!

**Comment by marcin-krystianc** [comment]
> The only issue I have is with another test case that the Consul in Golang has: https://github.com/hashicorp/consul/blob/v1.9.17/api/connect_ca_test.go#L12 . They use the testutil to create another client for testing and do not bootstrap connect in that test case, which is, however, not really possible with the current testing setup that consuldotnet has. Because that would require to create another test_config.json and run both the new and the old one when testing. Please correct me if I am wrong!

You are right, but we don't have to replicate that test. The test you've already added is sufficient. It tests that the endpoint works and returns desired data.

File: Consul.Test/AgentTest.cs [hunk 1]
```diff
@@ -1025,6 +1025,15 @@ public async Task Agent_Metrics()
             Assert.NotNull(agentMetrics.Response.Samples);
         }
 
+        [Fact]
+        public async Task Agent_CARoots()
+        {
+            var caRoots = await _client.Agent.GetCARoots();
+            Assert.NotEqual((ulong)0, caRoots.LastIndex);
+            Assert.Single(caRoots.Response.Roots);
+            Assert.Equal("11111111-2222-3333-4444-555555555555.consul", caRoots.Response.TrustDomain);
```
Comment (marcin-krystianc) [inline Consul.Test/AgentTest.cs:1034]: This test could test all fields in the response (at least for fields being non-null). That would assure us that all fields in the newly added structures are named correctly.

**Review by marcin-krystianc — COMMENTED** [review]
Generally, it looks good. Please add more test assertions so we know that all the fields are named correctly.

**Comment by IvanKolchanov** [comment]
> > The only issue I have is with another test case that the Consul in Golang has: https://github.com/hashicorp/consul/blob/v1.9.17/api/connect_ca_test.go#L12 . They use the testutil to create another client for testing and do not bootstrap connect in that test case, which is, however, not really possible with the current testing setup that consuldotnet has. Because that would require to create another test_config.json and run both the new and the old one when testing. Please correct me if I am wrong!
> 
> You are right, but we don't have to replicate that test. The test you've already added is sufficient. It tests that the endpoint works and returns desired data.

That's what I thought, just had to make sure! :)

File: Consul.Test/AgentTest.cs (hunk 1 above)
Comment (IvanKolchanov) [reply Consul.Test/AgentTest.cs]: Just added the test assertions, ready for further review!

**Comment by IvanKolchanov** [comment]
@marcin-krystianc I see that a few tests had failed because of an assertion error, expected that a bit, because I wasn't sure which kind of values can SerialNumber, CreateIndex and ModifyIndex take depending on conditions. PrivateKeyBits shouldn't be 0 though. Don't think that there is a way for me to initialize those values in config, except for PrivateKeyBits (https://developer.hashicorp.com/consul/docs/connect/ca/consul). Should I just delete the assert tests for SerialNumber, CreateIndex and ModifyIndex?
Thank you in advance!

**Comment by marcin-krystianc** [comment]
> @marcin-krystianc I see that a few tests had failed because of an assertion error, expected that a bit, because I wasn't sure which kind of values can SerialNumber, CreateIndex and ModifyIndex take depending on conditions. PrivateKeyBits shouldn't be 0 though. Don't think that there is a way for me to initialize those values in config, except for PrivateKeyBits (https://developer.hashicorp.com/consul/docs/connect/ca/consul). Should I just delete the assert tests for SerialNumber, CreateIndex and ModifyIndex? Thank you in advance!

It seem that the problem only occurs for version 1.6.10. So perhaps we can use `AgentVersion` variable (https://github.com/G-Research/consuldotnet/blob/219ecd42b53d7b0114aea540e5fb9a0fbce1a56e/Consul.Test/BaseFixture.cs#L15) and assert those problematic fields only when the version is equal or greater than `1.7.0` ?

File: Consul.Test/AgentTest.cs [hunk 2]
```diff
@@ -1025,6 +1025,34 @@ public async Task Agent_Metrics()
...
+            Assert.NotNull(root.Name);
+            Assert.NotNull(root.SigningKeyID);
+            Assert.NotNull(root.ExternalTrustDomain);
+            Assert.NotNull(root.NotBefore);
+            Assert.NotNull(root.NotAfter);
+            Assert.NotNull(root.RootCert);
+            Assert.Null(root.IntermediateCerts);
+            Assert.True(root.Active);
+            Assert.NotNull(root.PrivateKeyType);
+            Assert.NotEqual(0, root.PrivateKeyBits);
```
Comment (marcin-krystianc) [inline Consul.Test/AgentTest.cs:1047]: I think this one needs to go inside the `if (AgentVersion >= SemanticVersion.Parse("1.7.0"))` block.

Comment (IvanKolchanov) [reply Consul.Test/AgentTest.cs]: Yeah, a bit surprising, but should be working now, fixed it!

**Review by marcin-krystianc — APPROVED** [review]
I think now it is perfect, thank you!

**Pull request merged**